plugin/trino-redshift/src/main/java/io/trino/plugin/redshift/RedshiftUnloadSplitManager.java

```diff
@@ -0,0 +1,200 @@
+    {
+        if (table instanceof JdbcProcedureHandle) {
+            return jdbcSplitManager.getSplits(transaction, session, table, dynamicFilter, constraint);
+        }
+        if (!useUnload(session)) {
+            return new FixedSplitSource(new JdbcSplit(Optional.empty()));
+        }
+        JdbcTableHandle jdbcTableHandle = dynamicFilteringEnabled(session) ? ((JdbcTableHandle) table).intersectedWithConstraint(dynamicFilter.getCurrentPredicate()) : (JdbcTableHandle) table;
+        List<JdbcColumnHandle> columns = jdbcTableHandle.getColumns()
+                .orElseGet(() -> jdbcClient.getColumns(
+                        session,
+                        jdbcTableHandle.getRequiredNamedRelation().getSchemaTableName(),
+                        jdbcTableHandle.getRequiredNamedRelation().getRemoteTableName()));
+
+        ConnectorSplitSource fallbackSplitSource = new FixedSplitSource(new JdbcSplit(Optional.empty()));
+
+        // Fallback to jdbc for no table columns in projection
+        if (((JdbcTableHandle) table).getColumns().isPresent() && ((JdbcTableHandle) table).getColumns().get().isEmpty()) {
+            return fallbackSplitSource;
+        }
+        if (containsUnsupportedType(columns)) {
+            return fallbackSplitSource;
+        }
+        if (((JdbcTableHandle) table).getLimit().isPresent()) {
+            return fallbackSplitSource;
+        }
+        if (containsFilterConditionWithDecimalTypeColumn(table)) {
+            return fallbackSplitSource;
+        }
+        Connection connection;
+        PreparedStatement statement;
+        try {
+            connection = jdbcClient.getConnection(session);
+            statement = buildUnloadSql(session, connection, jdbcTableHandle, columns);
+        }
+        catch (SQLException e) {
+            throw new RuntimeException(e);
+        }
+        return new RedshiftUnloadSplitSource(executor, connection, statement, fallbackSplitSource);
+    }
+
+    private PreparedStatement buildUnloadSql(ConnectorSession session, Connection connection, JdbcTableHandle table, List<JdbcColumnHandle> columns)
+            throws SQLException
+    {
+        PreparedQuery preparedQuery = jdbcClient.prepareQuery(session, table, Optional.empty(), columns, ImmutableMap.of());
+
+        PreparedStatement openTelemetryPreparedStatement = queryBuilder.prepareStatement(jdbcClient, session, connection, preparedQuery, Optional.of(columns.size()));
+        RedshiftPreparedStatement redshiftPreparedStatement;
+        CachedQuery redshiftPreparedQuery;
+        ParameterList preparedParameters;
+        try {
+            Field preparedStatementField = OpenTelemetryStatement.class.getDeclaredField("delegate");
+            preparedStatementField.setAccessible(true);
+            redshiftPreparedStatement = (RedshiftPreparedStatement) preparedStatementField.get(openTelemetryPreparedStatement);
```

Comment:
Can we use `PreparedStatement#unwrap` and pass argument as `RedshiftPreparedStatement.class` to fetch the underlying impl.

Reply:
Good to know. Changed it.